src/ipc/ipc4/handler.c

```diff
@@ -996,6 +996,36 @@ static int ipc4_set_get_config_module_instance(struct ipc4_message_request *ipc4
 	return ret;
 }
 
+__cold static void ipc4_prepare_for_kcontrol_get(struct comp_dev *dev, uint8_t param_id,
+						 char *data_out, uint32_t data_size)
+{
+	const char *hostbox;
+
+#if CONFIG_LIBRARY
+	hostbox = (const char *)ipc_get()->comp_data + sizeof(struct ipc4_module_large_config);
+#else
+	hostbox = (const char *)MAILBOX_HOSTBOX_BASE;
+#endif
+
+	assert_can_be_cold();
+
+	switch (param_id) {
+	case SOF_IPC4_SWITCH_CONTROL_PARAM_ID:
+	case SOF_IPC4_ENUM_CONTROL_PARAM_ID:
+	case SOF_IPC4_BYTES_CONTROL_PARAM_ID:
+		/* We have payload in hostbox */
+		dcache_invalidate_region((__sparse_force void __sparse_cache *)MAILBOX_HOSTBOX_BASE,
+					 data_size);
+
+		/* Copy the control payload header from inbox to outbox */
+		memcpy_s(data_out, data_size, hostbox,
+			 sizeof(struct sof_ipc4_control_msg_payload));
+		break;
+	default:
+		break;
```

Comment:
would we more often hit one of the three `case`s above or end up in `default` case here? If `default` is rare, we could unite the `dcache_invalidate_region()` calls above and in line 1175 to put it directly below line 1170. This is optional and can also be an incremental PR. But if we usually take the `default` path, then maybe it isn't a good idea.

Reply:
This will be highly topology dependent and we will only hit these where there is TDFB used (and it is used by user space as well) or we have sound dose enabled and actively used.

During runtime we are likely not even receiving GET_LARGE_CONFIG messages at all. I cannot say which will be the most frequent use in the future.